src/main/java/com/google/devtools/build/lib/syntax/CallUtils.java

```diff
@@ -41,11 +39,12 @@ private static CacheValue getCacheValue(Class<?> cls, StarlarkSemantics semantic
     if (cls == String.class) {
       cls = StringModule.class;
     }
-    try {
-      return cache.get(new Key(cls, semantics));
-    } catch (ExecutionException ex) {
-      throw new IllegalStateException("cache error", ex);
+    Key key = new Key(cls, semantics);
+    CacheValue cacheValue = cache.get(key);
```

Comment:
Why call get then computeIfAbsent? Does this reduce contention? If so, by how much? State in a comment. If not, simplify.

Reply:
`computeIfAbsent` without `get` makes it 1% slower or something (hard to say precisely because of high variance).

For reasons I don't understand `computeIfAbsent` always acquires lock even if it is a hit (at least in JDK on my laptop).

Reply:
> computeIfAbsent without get makes it 1% slower or something (hard to say precisely because of high variance).

Thanks. Could you note this in a comment?

> For reasons I don't understand computeIfAbsent always acquires lock even if it is a hit (at least in JDK on my laptop).

Well, of course it has to, because it guarantees atomicity (which is stronger than the general Map.computeIfAbsent contract).

Reply:
Added a comment.

Reply:
> Well, of course it has to, because it guarantees atomicity (which is stronger than the general Map.computeIfAbsent contract).

That still doesn't make sense to me. It seems it would get the same guarantee by doing a normal get() first and then doing exactly what it's doing now.

From the doc:
> The entire method invocation is performed atomically, so the function is applied at most once per key.

The second part of that (the function only being applied at most once) would be fine with a speculative get(). I suppose you could read the first part of that as not allowing a speculative get(), but whether it's there or not doesn't change the observable behavior or guarantees (afaict).

Reply:
Ah, and openjdk considered it a bug that computeIfAbsent() was locking even when present. They've fixed it looks like in jdk9: https://bugs.java.com/bugdatabase/view_bug.do?bug_id=8161372